src/NuGet.Services.AzureSearch/Owners2AzureSearch/PackageIdToOwnersBuilder.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.Logging;
+
+namespace NuGet.Services.AzureSearch.Owners2AzureSearch
+{
+    public class PackageIdToOwnersBuilder
+    {
+        private readonly ILogger _logger;
+        private int _addCount;
+        private readonly Dictionary<string, string> _usernameToUsername;
+        private readonly SortedDictionary<string, SortedSet<string>> _result;
+
+        public PackageIdToOwnersBuilder(ILogger logger)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _addCount = 0;
+            _usernameToUsername = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            _result = new SortedDictionary<string, SortedSet<string>>(StringComparer.OrdinalIgnoreCase);
+        }
+
+        public void Add(string id, IEnumerable<string> owners)
+        {
+            foreach (var username in owners)
+            {
+                Add(id, username);
+            }
+        }
+
+        public void Add(string id, string username)
+        {
+            _addCount++;
+            if (_addCount % 10000 == 0)
+            {
+                _logger.LogInformation("{AddCount} ownership records have been added so far.", _addCount);
+            }
+
+            // Use a single instance of each username string.
+            if (!_usernameToUsername.TryGetValue(username, out var existingUsername))
```

Comment:
Why do you use a dictionary here instead of `string.Intern`?

Reply:
`string.Intern` never gets GC'd. First thing I learned on NuGet team thanks to @yishaigalatzer 👍.

Reply:
I’d be more concerned by the global lock used in `string.Intern`’s implementation. Regardless, the set of owners on nuget.org is relatively stable :)

Reply:
Concerning the lock, that's their implementation detail and would not be a problem given we're doing this pseudo-interning serially.

Concerning the set of owners being stable, this is true but the job does not need to hold onto this memory why it is sleeping. Current design is to run every 24 hours so a lot of the time the memory will be used when it is not needed.

Reply:
Ah that's a good point 👍